File: Remove_Duplicates_from_Sorted_List.cc
/*
Given a sorted linked list, delete all duplicates such that each element appear only once.

For example,
Given 1->1->2, return 1->2.
Given 1->1->2->3->3, return 1->2->3.
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution 
{
public:
    ListNode *deleteDuplicates(ListNode *head) 
    {
        if (head == NULL || head->next == NULL)
            return head;
            
        ListNode* pPre = head;
        ListNode* pCur = head->next;
        while (pCur != NULL) 
        {
            if (pPre->val == pCur->val)
            {
                pPre->next = pCur->next;
                delete pCur;
            }
            else
            {
                pPre->next = pCur;
                pPre = pPre->next;
            }
            
            pCur = pCur->next;
        }
        
        return head;
    }
};

File: go/remove-duplicates-from-sorted-list/remove-duplicates-from-sorted-list.go
package remove_duplicates_from_sorted_list

import (
	"../util"
)

/*
83. 删除排序链表中的重复元素

给定一个排序链表，删除所有重复的元素，使得每个元素只出现一次。

示例 1:

输入: 1->1->2
输出: 1->2
示例 2:

输入: 1->1->2->3->3
输出: 1->2->3
*/

type ListNode = util.ListNode

func deleteDuplicates(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	cur, next := head, head.Next

	for next != nil {
		if next.Val == cur.Val {
			cur.Next = next.Next // 删除next
			// cur 不动，next前移继续判断
			next = next.Next
		} else { // 不等，同时前移
			cur = next
			next = next.Next
		}
	}

	return head // head不会变
}
